### PR TITLE
Moving+editing events

### LINKED ISSUE #81: editing + Moving appointments in time
All properties that were specified during creation (#27 ) should be editable via a separate form.

Moving one single appointment should not move the whole appointment series, but only this specific one. (Maybe it would be a nice idea to additionally provide a "move series button" which achieves the other).

### PR DESCRIPTION
**Pull request by polodasi123**
<comment on what this PR does>

closes #81

---

### Implementation

- [x] Does this code change accomplish what it is supposed to do?
- [ ] Can this solution be simplified without unreasonable effort
- [ ] Is the code modular enough?
- [x] Does similar functionality already exist in the codebase? If yes, why isn’t it reused?
- [x] Single Responsibility Principle: Does any class serve only one purpose? If not, it should probably split up in multiple classes.
- [x] Interface Segregation: If interface implemented, is it fully implemented?
- [x] Dependency Injection: Are central object instantiation processes managed by a single context?
- [x] If any functionality is missing, is it marked with a comment that starts with TODO:?
- [x] Do you see any potential to improve the performance of the code significantly? 

### Logic Errors and Bugs

- [x] Can you think of any use case in which the code does not behave as intended?
- [x] Can you think of any inputs or external events that could break the code?
- [ ] Are edge cases covered by tests?

### Error Handling and Logging

- [x] Is error handling done the correct way?
- [x] Should any logging or debugging information be added or removed?
- [ ] Are error messages user-friendly?

### Dependencies (internal and external)

- [x] Were updates to documentation, configuration, or readme files made as required by this change?
- [x] Is the API.md up-to-date?
- [x] Are the enumerations between frontend and backend synchronized?

### Security and Data Privacy

- [x] Does the code introduce any security vulnerabilities?
- [x] Is (user) input validated, sanitized, and escaped to prevent security attacks such as cross-site scripting or SQL injection?
- [x] Is sensitive data like user data handled and stored securely?
- [x] Is the right encryption used?
- [x] Is data retrieved from external APIs or libraries checked for security issues?
- [x] Were any secrets committed?

### Testing and Testability (later)

- [ ] Have automated tests been added, or have related ones been updated to cover the change?
- [ ] Do the existing tests reasonably cover the code change?
- [ ] Are there some test cases, input or edge cases that should be tested in addition?

### Usability and Accessibility

- [x] Is the proposed solution (UI) accessible (e.g. colorblindness, ...)?
- [ ] Is the API/UI intuitive to use?
- [x] Was the app tested on another device?

### Readability

- [ ] Is the code easy to understand?
- [x] Which parts were confusing to you and why?
- [x] Can the readability of the code be improved by smaller methods/different names?
- [x] Is the code located in the right file/folder?
- [x] Do you think certain methods should be restructured to have a more intuitive control flow?
- [x] Are there redundant or outdated comments?
- [x] Could some comments convey the message better?
- [x] Could some comments be removed by making the code itself more readable?

### CODE REVIEW

File: VLA-Frontend/src/components/calendar/EventDetails/MoveEventDialog.tsx [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    return `${event.dateISO}T09:00`;
+  };
+
+  const getCurrentEndDateTime = () => {
+    if (event.displayedEndTime) {
+      // displayedEndTime ist bereits im Format "HH:MM"
+      return `${event.dateISO}T${event.displayedEndTime}`;
+    }
+    return `${event.dateISO}T10:00`;
+  };
```
Comment (MarcoLeipnitz) [inline VLA-Frontend/src/components/calendar/EventDetails/MoveEventDialog.tsx]: Utility functions

File: VLA-Frontend/src/components/calendar/EventForm/AddLectureForm.tsx [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+  onSubmit: (lecture: Lecture) => void;
+  onCancel: () => void;
+  people?: Person[];
+  onAddPerson?: (person: Person) => void;
+};
+
+/**
+ * ============================================================================
+ * NEUE KOMPONENTE: AddLectureForm
+ * ============================================================================
```
Comment (MarcoLeipnitz) [inline VLA-Frontend/src/components/calendar/EventForm/AddLectureForm.tsx]: I do not think the NEUE KOMPONENTE part is neccessary

File: VLA-Frontend/src/components/calendar/CalendarTypes.ts [hunk 3]
```diff
@@ -37,16 +34,25 @@ export type CalendarEvent = {
...
 export type CalendarEventsByDateISO = Record<string, CalendarEvent[]>;
 
-
+export type Person = {
+  id: string;
+  name: string;
+  email?: string;
+  role?: string;
+  notes?: string;
+};
```
Comment (MarcoLeipnitz) [inline VLA-Frontend/src/components/calendar/CalendarTypes.ts]: Please look at #104, I think my PR should be merged to avoid such types

File: VLA-Frontend/src/hooks/usePeople.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
+import { useState } from "react";
+import type {Person} from "@/lib/databaseTypes";
+
+/**
+ * usePeople manages the list of people that can be assigned to events.
+ */
+export function usePeople() {
```
Comment (MarcoLeipnitz) [inline VLA-Frontend/src/hooks/usePeople.ts]: Consider looking into the new SseConnection hook(s) (which might come handy in this context.

File: VLA-Frontend/src/hooks/useLectures.ts
Comment (MarcoLeipnitz) [inline VLA-Frontend/src/hooks/useLectures.ts]: Consider looking into the SseConnection hooks from main

File: VLA-Frontend/src/components/calendar/EventDetails/MoveEventDialog.tsx [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+/**
+ * MoveEventDialog allows the user to move an event to a new date/time
+ */
+export default function MoveEventDialog({ 
+  event, 
+  onMove, 
+  onCancel, 
+  isSeries, 
+}: MoveEventDialogProps) {
+  const durationChanged = useRef(false);
```
Comment (MarcoLeipnitz) [inline VLA-Frontend/src/components/calendar/EventDetails/MoveEventDialog.tsx]: @JohaJung, You asked about checking state in the next invocation of a function yesterday.
I think useRef is what you are looking for.

File: VLA-Frontend/src/components/calendar/EventDetails/MoveEventDialog.tsx [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+          <div className="cv-formGroup">
+            <label htmlFor="newStartDateTime" className="cv-formLabel">
+              Neuer Start (Datum & Uhrzeit) *
+            </label>
+            <input
+              id="newStartDateTime"
+              type="datetime-local"
+              className="cv-formInput"
+              value={newStartDateTime.toISOString()}
```
[the remark this answers is not in the document]
Comment (MarcoLeipnitz) [reply VLA-Frontend/src/components/calendar/EventDetails/MoveEventDialog.tsx]: Below as well

File: VLA-Frontend/src/components/calendar/EventForm/AddLectureSection.tsx [hunk 7]
```diff
@@ -1,12 +1,14 @@
...
-import type { Lecture } from "../CalendarTypes";
-import ColorPicker from "../ColorPicker";
+import AddLectureForm from "./AddLectureForm";
+import type {Lecture, Person} from "@/lib/databaseTypes";
 
 type AddLectureSectionProps = {
   lectures: Lecture[];
-  selectedLectureId: string;
-  onLectureChange: (lectureId: string) => void;
+  selectedLecture: Lecture | undefined;
```
Comment (MarcoLeipnitz) [inline VLA-Frontend/src/components/calendar/EventForm/AddLectureSection.tsx]: Why not just use selectedLecture?: Lecture

File: VLA-Frontend/src/components/calendar/eventUtils.ts [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    };
+  });
+}
+
+
+/**
+ * Get event status.
+ */
+export function getEventStatus(_event: Appointment) {
+  // TODO: implement, extract experiments? What was this EventStatus type thought for?
```
Comment (MarcoLeipnitz) [inline VLA-Frontend/src/components/calendar/eventUtils.ts]: See Calendartypes.ts
I suppose this should represent the ExperimentationPreparationStatus, so it should probably be changed

**Comment by MarcoLeipnitz** [comment]
I cannot set a starting time without it disappearing when I type it in.

File: VLA-Frontend/src/components/calendar/eventUtils.ts (hunk 8 above)
Comment (JohaJung) [reply VLA-Frontend/src/components/calendar/eventUtils.ts]: This also a point where I avoided to change logic, just did refactoring.

File: VLA-Frontend/src/components/calendar/EventForm/AddCategorySection.tsx [hunk 9]
```diff
@@ -27,7 +27,11 @@ export default function AddCategorySection({
   function handleAdd() {
     if (newCategoryName.trim() === "") return;
 
-    onAddCategory(newCategoryName.trim());
+    const newCategory: AppointmentCategory = {
```
[the remark this answers is not in the document]
Comment (JohaJung) [reply VLA-Frontend/src/components/calendar/EventForm/AddCategorySection.tsx]: Because this is not a function, but an object :grin:

File: VLA-Frontend/src/components/calendar/EventDetails/MoveEventDialog.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (JohaJung) [reply VLA-Frontend/src/components/calendar/EventDetails/MoveEventDialog.tsx]: fixed, and using the central TimeRangeInput component now.

File: VLA-Frontend/src/hooks/useLectures.ts
Comment (JohaJung) [reply VLA-Frontend/src/hooks/useLectures.ts]: In this PR? What is the advantage of that?

**Comment by JohaJung** [comment]
This works (again), now. I hopefully fixed all bugs introduced during the refactoring.

Of course, the code could be nicer, but I (or Marco) want to start integrating #103 soon.

@polodasi123 möchtest du einmal testen ob sich hier irgendwas am Verhalten unerwartet geändert hat?

File: VLA-Frontend/src/components/calendar/EventDetails/EventEditForm.tsx [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+      start: startDateTime,
+      end: endDateTime,
+      notes: notes.trim(),
+      series: updatedSeries,
+    };
+
+    onSave(updatedEvent);
+  };
+
+  const hasTitle = (title.trim() !== "") || lecture;
```
Comment (MarcoLeipnitz) [inline VLA-Frontend/src/components/calendar/EventDetails/EventEditForm.tsx]: Please make the boolean on the right explicit.
While this is valid syntax and semantically correct, I think it is weird to read boolean || string

File: VLA-Frontend/src/components/calendar/EventForm/AddLectureForm.tsx [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+  };
+
+  const handleAddPerson = (person: Person) => {
+    onAddPerson?.(person);
+    setSelectedPeople([...selectedPeople, person.id]);
+  };
+
+  const isValid = lectureName.trim() !== "";
+
+  const madalContent = (
```
[the remark this answers is not in the document]
Comment (MarcoLeipnitz) [reply VLA-Frontend/src/components/calendar/EventForm/AddLectureForm.tsx]: This is actually something that should be changed before merging

File: VLA-Frontend/src/components/calendar/eventUtils.ts (hunk 8 above)
Comment (MarcoLeipnitz) [reply VLA-Frontend/src/components/calendar/eventUtils.ts]: Remove this function and replace it with a proper enum

File: VLA-Frontend/src/hooks/useEvents.ts [hunk 12]
```diff
@@ -1,81 +1,179 @@
...
+    if (!oldEvent) {
+      Logger.error("Event not found");
+      return;
+    }
+    // create new appointment series when moving out of existing series
+    let newSeries: AppointmentSeries | undefined;
+    if (allEvents.filter(e => e.series.id === oldEvent.series.id).length > 1) {
+      newSeries = {
+        ...oldEvent.series,
+        id: -Date.now(),  // negative ID signals a not-yet-created entity
```
Comment (MarcoLeipnitz) [inline VLA-Frontend/src/hooks/useEvents.ts]: Negative date creation

File: VLA-Frontend/src/hooks/useEvents.ts [hunk 13]
```diff
@@ -1,81 +1,179 @@
...
+    let dummyId = -Date.now() - 1; // negative ID signals a not-yet-created entity
+
+    // create series of recurring events by looping day by day
+    if (formData.recurrence && formData.recurrence.weekdays.length > 0 && formData.recurrence.endDay) {
+      // create appointment series, one for each recurrence weekday
+      const seriesByWeekday = new Map<Weekday, AppointmentSeries>;
+      formData.recurrence.weekdays.forEach(
+        (day) => {
+          seriesByWeekday.set(day, {
+            id: --dummyId,
```
Comment (MarcoLeipnitz) [inline VLA-Frontend/src/hooks/useEvents.ts]: This seems unintentional?
Date problem all over

**Comment by JohaJung** [comment]
> Persons in series cannot be changed after series is created.

This is true, we should add a LectureEditForm (since the Persons are at the Lecture level, not at the series). Later, in a new PR.

> Personalized notes in series are not saved on exit

This got fixed now, notes can a) be edited directly at the evend and not when editing a whole series (since they may differ between different events in a series)

> When moving a series I do not quite get the start and end time. Why do they have to be on the same day? I thought that they would just change the time frame of the recurrence pattern

I added a hint text displaying a) the former time range and b) that all events will be move *correspondingly*. Do you think this is now understandable? @MarcoLeipnitz 

Screenshot

File: VLA-Frontend/src/components/calendar/EventDetails/EventEditForm.tsx [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+  lectures = [],
+  categories = [],
+  people = [],
+  onSave,
+  onCancel,
+  onAddCategory,
+  onAddPerson,
+  onAddLecture,
+}: EventEditFormProps) {
+  // TODO: proper copying, do not modify existing entity
```
[the remark this answers is not in the document]
Comment (JohaJung) [reply VLA-Frontend/src/components/calendar/EventDetails/EventEditForm.tsx]: Probably never, this is something we can care about when implementing front-backend sync soon.

**Comment by JohaJung** [comment]
> We need a PR for lecturer specific notes

No? This is implemented. Probably we should make it more prominent, together with the mail-to-all button

File: VLA-Frontend/src/components/calendar/EventDetails/PersonDetails.tsx [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+/**
+ * PersonDetails shows detailed information about a person in a modal.
+ */
+export default function PersonDetails({ person, onClose, onSaveNotes }: PersonDetailsProps) {
+  const [notes, setNotes] = useState(person.notes || "");
+ 
+  const handleSave = () => {
+    if (onSaveNotes) {
+    //TODO: Backend - PUT request to save notes
+      onSaveNotes(person.id, notes);
```
Comment (MarcoLeipnitz) [inline VLA-Frontend/src/components/calendar/EventDetails/PersonDetails.tsx]: I assume the saving status is not handled correctly by object holding the state of the note

File: VLA-Frontend/src/components/calendar/EventForm/AddCategorySection.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (MarcoLeipnitz) [reply VLA-Frontend/src/components/calendar/EventForm/AddCategorySection.tsx]: I did not see that

File: VLA-Frontend/src/hooks/useEvents.ts [hunk 16]
```diff
@@ -1,118 +1,202 @@
...
-            kind: formData.category,
-            status: formData.status,
-            shortTitle: formData.people.join(", "),
-            displayedStartTime: startTime,
-            displayedEndTime: endTime,
-            lectureId: formData.lectureId,
+            id: notSynchronisedId--,
+            series: seriesByWeekday.get(weekday)!,
+            start: currentDate,
+            end: new Date(currentDate.getTime() + duration),
```
Comment (MarcoLeipnitz) [inline VLA-Frontend/src/hooks/useEvents.ts]: I think there is a utils functions for that

Comment (JohaJung) [reply VLA-Frontend/src/hooks/useEvents.ts]: I disagree.
